src/parser/jobs/mnk/modules/TwinSnakes.tsx

```diff
@@ -0,0 +1,226 @@
+
+	// Clipping the duration, or dropping for more than TS itself
+	private earlySnakes: number = 0
+	private lateSnakes: number = 0
+
+	// Separate accounting from the window, to handle counting while TS is down
+	private gcdsSinceTS: number = 0
+
+	protected init() {
+		// Hook all GCDs so we can count GCDs in buff windows
+		this.addHook('cast', {by: 'player'}, this.onCast)
+
+		// This gets weird because, we don't wanna penalise if it was from FPF...
+		this.addHook('applybuff', {to: 'player', abilityId: STATUSES.TWIN_SNAKES.id}, this.onGain)
+		this.addHook('refreshbuff', {to: 'player', abilityId: STATUSES.TWIN_SNAKES.id}, this.onRefresh)
+		this.addHook('removebuff', {to: 'player', abilityId: STATUSES.TWIN_SNAKES.id}, this.onDrop)
+
+		this.addHook('complete', this.onComplete)
+	}
+
+	private onCast(event: CastEvent): void {
+		const action = getDataBy(ACTIONS, 'id', event.ability.guid) as TODO
+
+		if (!action) {
+			return
+		}
+
+		// Only include GCDs
+		if (action.onGcd) {
+			switch (action.id) {
+			// Ignore TS itself, plus Form Shift and maybe 6SS?
+			// We use gcdsSinceTS because we don't want to double count FPF
+			case (ACTIONS.TWIN_SNAKES.id):
+				if (this.twinSnake && !this.twinSnake.end && this.gcdsSinceTS < TWIN_SNAKES_CYCLE_LENGTH) {
+					this.earlySnakes++
+				}
+
+			// Ignore Form Shift, theres probably forced downtime so we expect TS to get weird anyway
+			case (ACTIONS.FORM_SHIFT):
+				break
+
+			// Count FPF, but check if it's a bad one
+			case (ACTIONS.FOUR_POINT_FURY.id):
+				if (!this.combatants.selected.hasStatus(STATUSES.TWIN_SNAKES.id)) {
+					this.failedFury++
+				}
+
+			// Verify the window isn't closed, and count the GCDs
+			default:
+				if (this.twinSnake && !this.twinSnake.end) {
+					this.twinSnake.casts.push(event)
+				}
+
+				this.gcdsSinceTS++
+			}
+		}
+	}
+
+	// Only happens from TS itself
+	// This might be better checking if the GCD before it was buffed but ehh
+	private onGain(event: BuffEvent): void {
+		if (this.twinSnake && this.twinSnake.end) {
+			const unbuffedGcds = this.gcdsSinceTS - this.twinSnake.gcds
+			const unbuffedTime = event.timestamp - this.twinSnake.end
+			// TODO: some kind of downtime check, maybe a warning for non-GL4
+			if (unbuffedGcds > 1 || unbuffedTime > TWIN_SNAKES_CYCLE_BUFFER) {
+				this.lateSnakes++
+			}
+		}
+
+		// Start a new window
+		this.twinSnake = new TwinState(event.timestamp)
+		this.gcdsSinceTS = 0
+	}
+
+	// Can be TS or FPF - just reset the GCD count
+	// This is ok even with the reduced duration from FPF provided user is in GL3/4
+	private onRefresh(): void {
+		this.gcdsSinceTS = 0
+	}
+
+	private onDrop(event: BuffEvent): void {
+		this.stopAndSave(event.timestamp)
+	}
+
+	private onComplete() {
+		// Close off the last window
+		this.stopAndSave(this.parser.fight.end_time)
+
+		// Calculate derped potency to early refreshes
+		const lostTruePotency = this.earlySnakes * (ACTIONS.TRUE_STRIKE.potency - ACTIONS.TWIN_SNAKES.potency)
```

Comment:
Is this necessary? We can just make a note of how many times they refreshed early, and I think that's easier to parse.

Reply:
I think it's useful, just because of how hard MNK hits so being able to quantify potency is good. The actual suggestion includes how many times an early refresh was done on top of the potency.